### PR TITLE
Added PlatformIO support and fixed vulnerabilities

### PR DESCRIPTION
**Pull request by attermann**
Added `platformio.ini` with build profiles targeting all of the currentlty supported boards.
Addressed some vulnerabilities exposed by the verbose build in PlatformIO, including:
- Declaring update_csma_p() before first use in update_airtime()
- Returning true from bt_setup_hw() when bt_ready is already true
- Broke-up `(++disp_page%pages)` in draw_disp_area() to avoid unintended result due to undefined behavior (https://stackoverflow.com/questions/4176328/what-are-sequence-points-and-how-do-they-relate-to-undefined-behavior)
- Changed variable `uint16_t pb` in update_airtime() to signed since a negative value is anticipated
- Changed memcpy into `dev_firmware_hash_target` in serialCallback() to use `DEV_HASH_LEN` instead of `DEV_SIG_LEN` which would result in buffer overrun
- Removed negative value check of `addr` from eeprom_write() since variable is unsigned
- Commented SD_MISO, SD_MOSI, and SD_CS consts for BOARD_RNODE_NG_21 which conflict with defines already present (with matching values) in espressif32, and which are unnused anyway due to `HAS_SD false`
- Changed `x <= y <= z` which does not match what's mathematically expected in C++ to `(x <= y) && (y <= z)` which is the intended comparison. (https://stackoverflow.com/questions/58358046/i-can-do-x-y-z-how-come-x-y-z-is-not-allowed-in-c)
- Added parentheses around `&` comparisons

### CODE REVIEW

**Comment by attermann** [comment]
@markqvist please see description of [ecf8792] for discussion of an issue with modem status flags. I don't have a good enough understanding of the code to know whether that's the correct fix so would appreciate your review.

File: RNode_Firmware.ino [hunk 1]
```diff
@@ -423,10 +423,19 @@ void add_airtime(uint16_t written) {
...
+  void update_csma_p() {
+      csma_p = (uint8_t)((1.0-(csma_p_min+(csma_p_max-csma_p_min)*csma_slope(airtime)))*255.0);
+}
+#endif
+
 void update_airtime() {
   #if MCU_VARIANT == MCU_ESP32
     uint16_t cb = current_airtime_bin();
-    uint16_t pb = cb-1; if (pb < 0) { pb = AIRTIME_BINS-1; }
+    int16_t pb = cb-1; if (pb < 0) { pb = AIRTIME_BINS-1; }
```
Comment (markqvist) [inline RNode_Firmware.ino:438]: You're right this is nonsense as it is ;) But I'd prefer:
```c
uint16_t pb = cb-1; if (cb-1 < 0) { pb = AIRTIME_BINS-1; }
```

File: RNode_Firmware.ino [hunk 2]
```diff
@@ -868,7 +877,7 @@ void serialCallback(uint8_t sbyte) {
           }
 
           if (frame_len == DEV_HASH_LEN) {
-            memcpy(dev_firmware_hash_target, cmdbuf, DEV_SIG_LEN);
+            memcpy(dev_firmware_hash_target, cmdbuf, DEV_HASH_LEN);
```
Comment (markqvist) [inline RNode_Firmware.ino:880]: Whoops. Pretty good to have that one fixed!

File: RNode_Firmware.ino [hunk 3]
```diff
@@ -935,12 +944,14 @@ void updateModemStatus() {
     portEXIT_CRITICAL(&update_lock);
   #endif
 
-  if (status & SIG_DETECT == SIG_DETECT) { stat_signal_detected = true; } else { stat_signal_detected = false; }
-  if (status & SIG_SYNCED == SIG_SYNCED) { stat_signal_synced = true; } else { stat_signal_synced = false; }
-  if (status & RX_ONGOING == RX_ONGOING) { stat_rx_ongoing = true; } else { stat_rx_ongoing = false; }
+  if ((status & SIG_DETECT) == SIG_DETECT) { stat_signal_detected = true; } else { stat_signal_detected = false; }
+  if ((status & SIG_SYNCED) == SIG_SYNCED) { stat_signal_synced = true; } else { stat_signal_synced = false; }
+  if ((status & RX_ONGOING) == RX_ONGOING) { stat_rx_ongoing = true; } else { stat_rx_ongoing = false; }
```
Comment (markqvist) [inline RNode_Firmware.ino:949]: And here I remember myself going on to someone about right-hand associativity...

Well spotted! Functionally, it didn't make much difference, luckily, since SIG_DETECT stays enabled during SIG_SYNCED too, but good to have it correct now.

File: RNode_Firmware.ino [hunk 4]
```diff
@@ -935,12 +944,14 @@ void updateModemStatus() {
...
 
-  if (stat_signal_detected || stat_signal_synced || stat_rx_ongoing) {
-    if (stat_rx_ongoing) {
+  //if (stat_signal_detected || stat_signal_synced || stat_rx_ongoing) {
+  if (stat_signal_detected || stat_signal_synced) {
+    //if (stat_rx_ongoing) {
+    if (stat_signal_detected) {
```
Comment (markqvist) [inline RNode_Firmware.ino:954]: Being slightly pedantic, but this needs to go back to what it was - if `stat_rx_ongoing` drops to false, it's because the modem was put in standby, and we need to deactivate DCD as well.

File: RNode_Firmware.ino [hunk 5]
```diff
@@ -935,12 +944,14 @@ void updateModemStatus() {
...
-  if (status & SIG_SYNCED == SIG_SYNCED) { stat_signal_synced = true; } else { stat_signal_synced = false; }
-  if (status & RX_ONGOING == RX_ONGOING) { stat_rx_ongoing = true; } else { stat_rx_ongoing = false; }
+  if ((status & SIG_DETECT) == SIG_DETECT) { stat_signal_detected = true; } else { stat_signal_detected = false; }
+  if ((status & SIG_SYNCED) == SIG_SYNCED) { stat_signal_synced = true; } else { stat_signal_synced = false; }
+  if ((status & RX_ONGOING) == RX_ONGOING) { stat_rx_ongoing = true; } else { stat_rx_ongoing = false; }
 
-  if (stat_signal_detected || stat_signal_synced || stat_rx_ongoing) {
-    if (stat_rx_ongoing) {
+  //if (stat_signal_detected || stat_signal_synced || stat_rx_ongoing) {
+  if (stat_signal_detected || stat_signal_synced) {
```
Comment (markqvist) [inline RNode_Firmware.ino:952]: Ok, but see next comment

File: RNode_Firmware.ino [hunk 6]
```diff
@@ -423,10 +423,19 @@ void add_airtime(uint16_t written) {
   #endif
 }
 
+#if MCU_VARIANT == MCU_ESP32
```
Comment (markqvist) [inline RNode_Firmware.ino:426]: Generally not a fan of just moving stuff around, but fair enough here.

File: Display.h
[the remark this answers is not in the document]
Comment (markqvist) [reply Display.h:1]: Not quite sure in what newfangled C(xx) this would not work, or be undefined, but please correct me if I'm wrong ;)

File: Utilities.h [hunk 7]
```diff
@@ -1108,7 +1108,7 @@ void eeprom_update(int mapped_addr, uint8_t byte) {
 }
 
 void eeprom_write(uint8_t addr, uint8_t byte) {
-	if (!eeprom_info_locked() && addr >= 0 && addr < EEPROM_RESERVED) {
+	if (!eeprom_info_locked() && addr < EEPROM_RESERVED) {
```
Comment (markqvist) [inline Utilities.h:1111]: No harm here really to also check that addr is > 0. I know it's a uint, but not checking for that rides on the typo implicitly.

**Comment by markqvist** [comment]
Hi @attermann, thanks a bunch for the PR! Some of these changes should be committed to a firmware release immediately. Some of them I am hesitant about, but please see the comments I wrote about that.

A few I am actually hesitant about on a conceptual level, including the addition of PlatformIO support. Not that I am against that in any way as such, but I do not use that system myself - at all. And I'm pretty sure it won't happen at any point too. This means I have no way to support, test and maintain it.

If it is included in the repo by default, people will expect it to work, and ask questions and raise issues if/when it doesn't. Realistically, I cannot maintain and support this, so I won't. Using PlatformIO instead of the current build system may also lead to various "incongruences for aesthetic reasons", where stuff is moved around purely to satisfy the compiler warning levels of somebody elses tastes.

Now, I know full well that this is NOT you, and that you know what you are doing, so I'm only saying this for perspective, but the last time someone tried to be helpful, coming from the perspective of compiler warnings, they submitted a completely untested PR, that *absolutely entirely broke and destroyed the whole darned firmware*. We can laugh about this now, but I only have so much time, and I really **don't** want to invite stuff like that - the firmware is deployed on thousands of devices, and needs a high level of functional assurance before being compiled and released.

So, bottom line is that I don't really now the best way forward. If we are to include PlatformIO support, somebody has to stand up and volunteer the responsibility to maintain it, and test it across all supported devices every time a new release is prepared. That's a bit much to ask, so I'm not gonna do it, but just leaving the possibility open, in case you feel like you have lots of time on your hands ;)

**Comment by markqvist** [comment]
Also, this PR needs a bit of discussing and adjusting here and there, but in about 48 hours, I will be seriously off the grid for a long time, and I really want to get your important fixes out and released in a firmware update before that happens.

For now I will manually merge those in, compile and test the firmware on devices and release it. Then there's amble time to figure out the best approach on the PlatformIO thing.

Again, thanks for the effort!

File: LoRa.cpp
Comment (markqvist) [inline LoRa.cpp:1]: Yup, being deep in Python-land, I missed that one when reviewing #36.

**Comment by attermann** [comment]
Some of your comments above are missing references to the actual change being discussed so I'm not sure how to respond. This may be because I re-forked your repo to re-organize my changes.

I have the latest after your cherry-picking and still have one issue for when you return.

```/Users/chad/Documents/PlatformIO/RNode_Firmware_orig/RNode_Firmware.ino:447:7: error: 'update_csma_p' was not declared in this scope```

Not sure why the compiler in Arduino CLI accepts this, but it seems my compiler is treating the function as pure-C and requiring that it be declared prior to use. That was the reason for my moving this function to precede where it's first called.
